[AIRFLOW-8057] [AIP-31]  Add @task decorator (#8962)
mistercrunch/Airflow: docs/concepts.rst

    op.dag is dag # True

.. _concepts:functional_dags:

> Functional DAGs

Reviewer: Random thought. Functional dags is perhaps not the right name for this whole feature. It's not like functional programming where a something is run without side-effects.

What would people think about renaming this to "Literate DAGs"? I'll take to the mailing list if people agree, just something doesn't quite sit right with me about "functional" today.

If no one else likes this/thinks functional is great then I'll not worry.

Reviewer: And we can discuss this _after_ this PR is merged if we want. It's not a blocker on merging the PR.

Author: That's fair. The Functional DAGs comes from the AIP title itself. I do understand the confusion, but not sure if literate is better. 

Ideally though, having XComArg and all this, it should allow DAGs to be written without as much side effects. You will be able to paramatrize an operator such that it is indeed more functional (takes A and B, outputs C) instead of current approach where you mostly generate artifacts that depend on Airflow context and such they are side effects.

Not sure if the above makes sense. In my mind, it makes it more functional (less prone to side effects) as you make operators resemble more functions, and XComArgs map to function parameters and function outputs.

Reviewer 2: Does this need any name? Can't we just say that this is a way how users can define DAGs? Personally I would encourage users to use this way instead of using XComs. WDYT?

Author: Fair. Main issue is that it does introduce a new syntax so for backwards compatible issues I would add a name to this.

We can switch all documentation to use this style on 2.0 and therefore remove this specific section (or collapse it)

Reviewer 2: @casassg do we have any backwards issues other than metaclass?

Author: That's true. Mostly saying as a full refactor of user docs on how to write DAGs may be more adequate for Airflow 2.0 than 1.10. But happy to reconsider this. If we go down that path, we should do it in a separate PR though. This PR is already become large enough

Reviewer 2: Definitely! We should make an issue for that, discuss and then adjust the docs in a follow-up PR

Author: Created #9415

    op.dag is dag # True

.. _concepts:functional_dags:

Functional DAGs
---------------

> *Added in Airflow 1.10.11*

Reviewer: ```suggestion
```

Let's not put this in until we're _sure_ it's going to make it in to the release. Sorry.

I would _love_ to have this in 1.10.11, but it might not make it in. We can add this in a separate PR once we've know for sure if it makes it in or not

DAGs can be defined using functional abstractions. Outputs and inputs are sent between tasks using
:ref:`XComs <concepts:xcom>` values. In addition, you can wrap functions as tasks using the
:ref:`task decorator <concepts:task_decorator>`. Dependencies are automatically inferred from

> the message dependencies.

Reviewer: "from the message dependencies" isn't quite clear - what it means. (What's a message? _I_ know what you mean, but we need to reword this)

Author: "Airflow will also automatically add dependencies between tasks to ensure that XCom messages are available when operators are executed."

wdyt?

Reviewer: that `XCom values`, then SGTM.

Example DAG with functional abstraction

.. code:: python

  with DAG(
      'send_server_ip', default_args=default_args, schedule_interval=None
  ) as dag:

    # Using default connection as it's set to httpbin.org by default
    get_ip = SimpleHttpOperator(
        task_id='get_ip', endpoint='get', method='GET', xcom_push=True
    )

>     @dag.task(multiple_outputs=True)

Reviewer: For the basic example I think it might be clearer if we didn't use "multiple_outputs"

Author: I rather do a more complete E2E tutorial here. We may want to switch the Hello World example on the tutorial page with an even simpler example. Thoughts?

Author: created #9041 as a follow up task

>     def prepare_email(raw_json: str) -> str:

Reviewer: This type hint is wrong.

```suggestion
    def prepare_email(raw_json: str) -> dict:
```

Or `Dict[str,str]` if you want to be more precise.

      external_ip = json.loads(raw_json)['origin']
      return {
        'subject':f'Server connected from {external_ip}',

>         'body': f'Seems like today your server executing Airflow is connected from the external IP {external_ip}<br>'

Reviewer 3: Unfortunately we can't use f-Strings, because we still support Python 2.7 in the 1.10.x series.

Reviewer 3: So if we want this in 1.10.11 we need to be very careful :/

Reviewer 2: The question is do we want to backport it?

Reviewer 3: Or we could use [future-fstrings](https://github.com/asottile/future-fstrings). We already have it as dep of a [dep](https://github.com/apache/airflow/search?q=future-fstrings&unscoped_q=future-fstrings) and add `# -*- coding: future_fstrings -*-`.

Author: Given this is an example DAG on a document, I don't think it really matters that much. If users want to use it on python 2 (which btw is already deprecated) they can change it to format. But I think we should adapt to use py3 examples better.

Reviewer 3: There are more cases where you are using F-Strings https://github.com/apache/airflow/pull/8962#discussion_r431670798

Reviewer 2: f-strings are not a blocker and we are using them in many places

      }

>     email_info = prepare_email(get_ip.output)

Reviewer: Do we still need this `.output`? The AIP didn't have that, and its much "nicer" without it.

Author: The AIP used `__call__` instead which was a bit more verbose and cofusing in my opinion. I rather keep `.output` as it allows for more composable DAGs.

Reviewer: Can you explain what you mean about "more composable DAGs"? I don't understand what you mean.

The AIP <https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-31%3A+Airflow+functional+DAG+definition> has this:

```python
  ip_info = get_ip()
  subject = email_subject_generator(ip_info)
  body = email_body_generator(ip_info)
  send_email(subject=subject, html_content=body)
```

So this PR doesn't implement what we voted on.

Author: The `output` part was implemented already as part of a previous PR. This PR focuses on `@task` only. 

The main difference between the proposed approach vs current approach is that in the current approach  we do not need to `call` normal operators. This should reduce complexity on creating functional DAGs. 

In addition, existing operators may have templated fields that are mandatory (example `subject` is required in `EmailOperator` initialization). This basically means that we need to overwrite them on `call` which is a bit confusing (see example below or example on AIP)

The `output` is basically a way to access XComArg for non callable operators. This allows a more smooth definition of functional DAGs between normal operators and `@task` operators.

Old example:
```python
    get_ip = SimpleHttpOperator(
        task_id='get_ip', endpoint='get', method='GET', xcom_push=True
    )

    @dag.task(multiple_outputs=True)
    def prepare_email(raw_json: str) -> Dict[str, str]:
      external_ip = json.loads(raw_json)['origin']
      return {
        'subject':f'Server connected from {external_ip}',
        'body': f'Seems like today your server executing Airflow is connected from the external IP {external_ip}<br>'
      }
    server_info = get_ip()
    email_info = prepare_email(server_info)
      
    send_email = EmailOperator(
        task_id='send_email',
        to='example@example.com',
        subject='',
        html_content=''
    )
    send_email(subject=email_info['subject'], html_content=email_info['body'])
```

New approach:
```python

    get_ip = SimpleHttpOperator(
        task_id='get_ip', endpoint='get', method='GET', xcom_push=True
    )

    @dag.task(multiple_outputs=True)
    def prepare_email(raw_json: str) -> Dict[str, str]:
      external_ip = json.loads(raw_json)['origin']
      return {
        'subject':f'Server connected from {external_ip}',
        'body': f'Seems like today your server executing Airflow is connected from the external IP {external_ip}<br>'
      }

    email_info = prepare_email(get_ip.output)

    send_email = EmailOperator(
        task_id='send_email',
        to='example@example.com',
        subject=email_info['subject'],
        html_content=email_info['body']
    )
```

I do believe the new approach is better. I'm happy to change the AIP and submit it for vote if that seems something that may be required.

Reviewer: Ahhh gotcha. I hadn't noticed/appreciated the difference between `ip_info = get_ip()` and `get_ip.output` and was merging them together in my head.

Reviewer: Could you update the AIP, and draft a short summary change to the list with lazy consensus (i.e. if no one objects within 48 hours we'll merge it) please?

Author: Sent message to mailing list: https://lists.apache.org/thread.html/rc40ea8171be80365c46ce29b4c8cfacc7567963c0df6af6056537c63%40%3Cdev.airflow.apache.org%3E

Author: Resolving after update passed in mailing list

We can say that task_1 is *upstream* of task_2, and conversely task_2 is *downstream* of task_1.
When a DAG Run is created, task_1 will start running and task_2 waits for task_1 to complete successfully before it may start.

.. _concepts:task_decorator:

Python task decorator
---------------------

Airflow ``task`` decorator converts any Python decorated function to a Python Airflow operator.
The decorated function can be called once to set the arguments and key arguments for operator execution.

.. code:: python

  with DAG('my_dag', start_date=datetime(2020, 5, 15)) as dag:

    @dag.task
    def hello_world():
      print('hello world!')

    # Also...

    from airflow.decorators import task

    @task
    def hello_name(name: str):
      print(f'hello {name}!')

    hello_name('Airflow users')

> Operator decorator captures returned values and sends them to the :ref:`XCom backend <concepts:xcom>`. By default, returned

Reviewer: Did you mean task, not operator?
```suggestion
Task decorator captures returned values and sends them to the :ref:`XCom backend <concepts:xcom>`. By default, returned
```

value is saved as a single XCom value. You can set ``multiple_outputs`` key argument to ``True`` to unroll dictionaries,
lists or tuples into seprate XCom values. This can be used with regular operators to create
:ref:`functional DAGs <concepts:functional_dags>`.

Calling a decorated function returns an ``XComArg`` instance. You can use it to set templated fields on downstream
operators.

> You can call a decorated function twice in a DAG. The decorated function will automatically generate unique a ``task_id``
> for each generated operator.

Reviewer: ```suggestion
You can call a decorated function more than once in a DAG. The decorated function will automatically generate unique 
a ``task_id`` for each generated operator.
```

.. code:: python

  with DAG('my_dag', start_date=datetime(2020, 5, 15)) as dag:

    @dag.task
    def update_user(user_id: int):
      ...

>     for user_id in user_ids:

Reviewer: Do we need to add a note here with caveats about "don't make the list dynamic/quickly changing"?

Author: Adding a note `Avoid generating this list dynamically to keep dag topology stable between DAG runs`

If we have dag serialization and dag versioning, that should be fine in the future though right? 👀

Author: Added comment.

We can say that task_1 is *upstream* of task_2, and conversely task_2 is *downstream* of task_1.
When a DAG Run is created, task_1 will start running and task_2 waits for task_1 to complete successfully before it may start.

.. _concepts:task_decorator:

Python task decorator
---------------------

> Airflow ``task`` decorator converts any Python decorated function to a Python Airflow operator.
> The decorated function can be called once to set the arguments and key arguments for operator execution.

Reviewer 4: ```suggestion
Airflow ``task`` decorator converts any Python function to an Airflow operator.
The decorated function can be called once to set the arguments and key arguments for operator execution.
```

.. code:: python

>   with DAG('my_dag', start_date=datetime(2020, 5, 15)) as dag:
> 
>     @dag.task
>     def hello_world():
>       print('hello world!')
> 
> 
>     # Also...
> 
>     from airflow.decorators import task
> 
>     @task
>     def hello_name(name: str):
>       print(f'hello {name}!')
> 
>     hello_name('Airflow users')

Reviewer 4: ```suggestion
  with DAG('my_dag', start_date=datetime(2020, 5, 15)) as dag:
      @dag.task
      def hello_world():
          print('hello world!')
  
  
      # Also...
      from airflow.decorators import task
  
  
      @task
      def hello_name(name: str):
          print(f'hello {name}!')
  
  
      hello_name('Airflow users')
```

Task decorator captures returned values and sends them to the :ref:`XCom backend <concepts:xcom>`. By default, returned
value is saved as a single XCom value. You can set ``multiple_outputs`` key argument to ``True`` to unroll dictionaries,
lists or tuples into seprate XCom values. This can be used with regular operators to create
:ref:`functional DAGs <concepts:functional_dags>`.

Calling a decorated function returns an ``XComArg`` instance. You can use it to set templated fields on downstream
operators.

> You can call a decorated function more than once in a DAG. The decorated function will automatically generate unique
> a ``task_id`` for each generated operator.

Reviewer 4: ```suggestion
You can call a decorated function more than once in a DAG. The decorated function will automatically generate
a unique ``task_id`` for each generated operator.
```

.. code:: python

  with DAG('my_dag', start_date=datetime(2020, 5, 15)) as dag:

    @dag.task
    def update_user(user_id: int):
      ...

>     # Avoid generating this list dynamically to keep dag topology stable between DAG runs

Reviewer 4: ```suggestion
    # Avoid generating this list dynamically to keep DAG topology stable between DAG runs
```

    for user_id in user_ids:

>       update_current(user_id)

Reviewer 3: ```suggestion
      update_user(user_id)
```

---------------------
*Added in Airflow 1.10.11*

Airflow ``task`` decorator converts any Python decorated function to a Python Airflow operator.
The decorated function can be called once to set the arguments and key arguments for operator execution.

.. code:: python

  with DAG('my_dag', start_date=datetime(2020, 5, 15)) as dag:

    @dag.task
    def hello_world():
      print('hello world!')

    # Also...

    from airflow.decorators import task

    @task
    def hello_name(name: str):
      print(f'hello {name}!')

    hello_name('Airflow users')

Operator decorator captures returned values and sends them to the :ref:`XCom backend <concepts:xcom>`. By default, returned
value is saved as a single XCom value. You can set ``multiple_outputs`` key argument to ``True`` to unroll dictionaries,
lists or tuples into seprate XCom values. This can be used with regular operators to create
:ref:`functional DAGs <concepts:functional_dags>`.

Calling a decorated function returns an ``XComArg`` instance. You can use it to set templated fields on downstream
operators.

If you call a decorated function twice in a DAG, it will error as it doesn't know what args to use.

> If you want to reuse decorated functions, use the copy method as follows:

Reviewer: I don't think this is the case anymore, right?

Author: updated.
